core/src/code/utils.rs

```diff
@@ -401,3 +431,153 @@ pub fn check_start_section(module: &Module) -> Result<(), CodeError> {
         Ok(())
     }
 }
+
+/// Calculates the data section size based on the number of OS pages used (see `GENERIC_OS_PAGE_SIZE`).
+pub fn get_data_section_bytes(module: &Module) -> Result<u32, CodeError> {
+    let Some(data_section) = module.data_section() else {
+        // No data section
+        return Ok(0);
+    };
+
+    let mut used_pages = BTreeSet::new();
+    for data_segment in data_section.entries() {
+        let data_segment_offset = data_segment
+            .offset()
+            .as_ref()
+            .and_then(get_init_expr_const_i32)
+            .ok_or(DataSectionError::Initialization)? as u32;
+        let data_segment_start = data_segment_offset / GENERIC_OS_PAGE_SIZE;
+
+        let data_segment_size = data_segment.value().len() as u32;
+        let data_segment_end = if data_segment_size == 0 {
+            data_segment_start
+        } else {
+            (data_segment_offset // We should use `offset` here and not `start`
+                .saturating_add(data_segment_size)
+                .saturating_add(GENERIC_OS_PAGE_SIZE - 1)) // Round up to the nearest whole number
+                / GENERIC_OS_PAGE_SIZE
+        };
+
+        for page in data_segment_start..data_segment_end {
+            used_pages.insert(page);
+        }
+    }
+
+    Ok(used_pages.len() as u32 * GENERIC_OS_PAGE_SIZE)
+}
+
+/// Calculates the amount of bytes in the global section will be initialized during module instantiation.
+pub fn get_global_section_bytes(module: &Module) -> Result<u32, CodeError> {
```

Comment:
I'll look into this

Reply:
It seems impossible because in different `get_<name>_section_size` functions, we use arguments `code` and `module`, which have non-overlapping lifetimes. This is because the serialize function consumes `module` in order to produce `code`.

This issue can be fixed by implementing `Serialize` for `AsRef<Module>` or something similar. However, I suggest doing this in a separate PR.

For reference:

```rust
        module = instrumentation_builder.instrument(module)?;

        // Use instrumented module to get section sizes.
        let data_section = utils::get_data_section_bytes(&module)?;
        let global_section = utils::get_global_section_bytes(&module)?;
        let table_section = utils::get_table_section_bytes(&module)?;
        let element_section = utils::get_element_section_bytes(&module)?;
        // HERE!
        let code = parity_wasm::elements::serialize(module) // <=== module consumed and code produced
            .map_err(CodecError::Encode)?; 

        // Use instrumented code to get section sizes.
        let CodeTypeSectionSizes {
            code_section,
            type_section,
        } = utils::get_code_type_sections_sizes(&code)?;
 ```

Reply:
Or did you mean to rewrite `get_<name>_section_size` functions in the way it's done with `get_code_type_sections_sizes` by simply re-parsing  `code`?

Reply:
After discussion decided to update function naming.